Added CMake dependencies on individual data files (#4496)
mfem/mfem: data/CMakeLists.txt

list(APPEND DATA_FILES
  amr-hex.mesh
  amr-quad.mesh
  ball-nurbs.mesh
  beam-hex-nurbs.mesh
  beam-hex.mesh
  beam-hex.vtk
  beam-quad-amr.mesh
  beam-quad-nurbs-sf.mesh
  beam-quad-nurbs.mesh
  beam-quad.mesh
  beam-quad.vtk
  beam-tet.mesh
  beam-tet.vtk
  beam-tri.mesh
  beam-tri.vtk
  beam-wedge.mesh
  beam-wedge.vtk
  compass.geo
  compass.mesh
  compass.msh
  cube-nurbs.mesh
  diag-segment-2d.mesh
  diag-segment-3d.mesh
  disc-nurbs.mesh
  escher-p2.mesh
  escher-p2.vtk
  escher-p3.mesh
  escher.mesh
  escher.vtk
  fichera-amr.mesh
  fichera-mixed-p2.mesh
  fichera-mixed-p2.vtk
  fichera-mixed.mesh
  fichera-q2.mesh
  fichera-q2.vtk
  fichera-q3.mesh
  fichera-quad-mixed.mesh
  fichera-quad.mesh
  fichera.mesh
  fichera.vtk
  hexagon.mesh
  inline-hex.mesh
  inline-pyramid.mesh
  inline-quad.mesh
  inline-segment.mesh
  inline-tet.mesh
  inline-tri.mesh

>   inline-wedge.mesh
>   klein-bottle.mesh
>   klein-donut.mesh
>   l-shape.mesh
>   llnl-p3.mesh
>   mobius-strip.mesh
>   octahedron.mesh
>   periodic-annulus-sector.geo
>   periodic-annulus-sector.msh
>   periodic-cube.geo
>   periodic-cube.mesh
>   periodic-cube.msh
>   periodic-hexagon.mesh
>   periodic-segment.mesh
>   periodic-square.geo
>   periodic-square.mesh
>   periodic-square.msh
>   periodic-torus-sector.geo
>   periodic-torus-sector.msh
>   pipe-nurbs-2d.mesh
>   pipe-nurbs-log.mesh
>   pipe-nurbs.mesh
>   ref-cube.mesh
>   ref-prism.mesh
>   ref-pyramid.mesh
>   ref-segment.mesh
>   ref-square.mesh
>   ref-tetrahedron.mesh
>   ref-triangle.mesh
>   rt-2d-p4-tri.mesh
>   rt-2d-q3.mesh
>   segment-nurbs.mesh
>   square-disc-nurbs-patch.mesh
>   square-disc-nurbs.mesh
>   square-disc-p2.mesh
>   square-disc-p2.vtk
>   square-disc-p3.mesh
>   square-disc-surf.mesh
>   square-disc.mesh
>   square-disc.vtk
>   square-mixed.mesh
>   square-nurbs-pw.mesh
>   square-nurbs.mesh
>   star-hilbert.mesh
>   star-mixed-p2.mesh
>   star-mixed-p2.vtk
>   star-mixed.mesh
>   star-q2.mesh
>   star-q2.vtk
>   star-q3.fms
>   star-q3.mesh
>   star-surf.mesh
>   star.mesh
>   star.vtk
>   tinyzoo-3d.mesh
>   toroid-hex.mesh
>   toroid-wedge.mesh
> )
> 
> foreach(DATA_FILE ${DATA_FILES})

Reviewer: ```suggestion
file(GLOB DATA_FILES ${CMAKE_CURRENT_SOURCE_DIR}/*)

foreach(DATA_PATH ${DATA_FILES})
  get_filename_component(DATA_FILE ${DATA_PATH} NAME)
```

Author: Right, I did try something similar but abandoned it because the reason you highlighted (which is why I think such loops are avoided in source directories in general).

Reviewer: I see, I've edited my suggestion above to trigger a reconfiguration if the `data` source directory is modified (which happens if you create a new file).

Author: Thanks @nmnobre , this looks promising: I'll test it out.

Author: This checks all the boxes, thanks @nmnobre! I have minor syntax things to tweak, so I implemented it in [9672921](https://github.com/mfem/mfem/pull/4496/commits/96729213ef37b69b6f8273e2d8e87f00e0ac4f20) instead of applying this suggestion directly. As such, I'll leave it to you to officially resolve this thread if there isn't more discussion.

> # set property that will initiate a configure step if data directory is modified

Reviewer 2: ```suggestion
# Copyright (c) 2010-2024, Lawrence Livermore National Security, LLC. Produced
# at the Lawrence Livermore National Laboratory. All Rights reserved. See files
# LICENSE and NOTICE for details. LLNL-CODE-806117.
#
# This file is part of the MFEM library. For more information and source code
# availability visit https://mfem.org.
#
# MFEM is free software; you can redistribute it and/or modify it under the
# terms of the BSD-3 license. We welcome feedback and contributions, see file
# CONTRIBUTING.md for details.

# If using `CMake >= 3.26` this functionality of this file can be replaced with `copy_directory_if_different`

# set property that will initiate a configure step if data directory is modified
```

Reviewer 2: Something like this is what I had in mind. Just a note for future developers if/when we enforce a higher version.

Author: Much appreciated (and thanks for adding the license statement)!

# Copyright (c) 2010-2024, Lawrence Livermore National Security, LLC. Produced
# at the Lawrence Livermore National Laboratory. All Rights reserved. See files
# LICENSE and NOTICE for details. LLNL-CODE-806117.
#
# This file is part of the MFEM library. For more information and source code
# availability visit https://mfem.org.
#
# MFEM is free software; you can redistribute it and/or modify it under the
# terms of the BSD-3 license. We welcome feedback and contributions, see file
# CONTRIBUTING.md for details.

> # If using `CMake >= 3.26` this functionality of this file can be replaced with `copy_directory_if_different`

Reviewer: ```suggestion
# If using `CMake >= 3.26` the functionality of this file can be replaced with `copy_directory_if_different`
```

> # If using `CMake >= 3.26` this functionality of this file can be replaced with `copy_directory_if_different`

Reviewer: ```suggestion
# If using `CMake >= 3.26`, the functionality of this file can be replaced with `copy_directory_if_different`
```

# set property that will initiate a configure step if data directory is modified
set_property(DIRECTORY APPEND
  PROPERTY CMAKE_CONFIGURE_DEPENDS ${CMAKE_CURRENT_SOURCE_DIR})

> set(COPIED_DATA_FILES)

Reviewer: ```suggestion
```

file(GLOB DATA_FILES ${CMAKE_CURRENT_SOURCE_DIR}/*)
foreach(DATA_FILE ${DATA_FILES})
  get_filename_component(DATA_FILENAME ${DATA_FILE} NAME)
  set(COPIED_DATA_FILE ${CMAKE_CURRENT_BINARY_DIR}/${DATA_FILENAME})
  add_custom_command(OUTPUT ${COPIED_DATA_FILE} DEPENDS ${DATA_FILE}
    COMMAND ${CMAKE_COMMAND} -E copy ${DATA_FILE} ${COPIED_DATA_FILE}
    COMMENT "")
  list(APPEND COPIED_DATA_FILES ${COPIED_DATA_FILE})
endforeach()

add_custom_target(copy_data DEPENDS ${COPIED_DATA_FILES}

>   COMMENT "Synching the data directory ...")

Reviewer: ```suggestion
  COMMENT "Syncing the data directory ...")
```

> # If using `CMake >= 3.26` this functionality of this file can be replaced with `copy_directory_if_different`
> 
> # set property that will initiate a configure step if data directory is modified
> set_property(DIRECTORY APPEND
>   PROPERTY CMAKE_CONFIGURE_DEPENDS ${CMAKE_CURRENT_SOURCE_DIR})
> 
> set(COPIED_DATA_FILES)
> file(GLOB DATA_FILES ${CMAKE_CURRENT_SOURCE_DIR}/*)
> foreach(DATA_FILE ${DATA_FILES})
>   get_filename_component(DATA_FILENAME ${DATA_FILE} NAME)
>   set(COPIED_DATA_FILE ${CMAKE_CURRENT_BINARY_DIR}/${DATA_FILENAME})
>   add_custom_command(OUTPUT ${COPIED_DATA_FILE} DEPENDS ${DATA_FILE}
>     COMMAND ${CMAKE_COMMAND} -E copy ${DATA_FILE} ${COPIED_DATA_FILE}
>     COMMENT "")
>   list(APPEND COPIED_DATA_FILES ${COPIED_DATA_FILE})
> endforeach()
> 
> add_custom_target(copy_data DEPENDS ${COPIED_DATA_FILES}
>   COMMENT "Synching the data directory ...")

Reviewer: ```suggestion
# If using `CMake >= 3.26`, the functionality of this file can be replaced with `copy_directory_if_different`

# set property that will initiate a configure step if data directory is modified
set_property(DIRECTORY APPEND
  PROPERTY CMAKE_CONFIGURE_DEPENDS ${CMAKE_CURRENT_SOURCE_DIR})

file(GLOB DATA_FILES ${CMAKE_CURRENT_SOURCE_DIR}/*)
foreach(DATA_FILE ${DATA_FILES})
  add_custom_command(OUTPUT ${DATA_FILE}_CMD
    COMMAND ${CMAKE_COMMAND} -E copy_if_different ${DATA_FILE} .
    COMMENT "")
  list(APPEND DATA_CMDS ${DATA_FILE}_CMD)
endforeach()

add_custom_target(copy_data DEPENDS ${DATA_CMDS}
  COMMENT "Syncing the data directory ...")
```

Reviewer: This does (1) to (4) and additionally copies if we modify a file in the build data directory.

>   COMMENT "Synching the data directory ...")

Reviewer: ```suggestion
  COMMENT "Syncing the data directory ...")
```
